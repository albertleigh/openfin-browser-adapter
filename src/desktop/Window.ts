import {BaseApiClass} from './Base';

import windowRegistry from '../helper/windowRegistry';

declare const window:any;

export interface WindowOptions {
    url?:string,
    name?:string,
}

export interface WindowBounds {
    width:number,
    height:number,
    top:number,
    left:number,
}

export class Window extends BaseApiClass{

    static _name:string = null;

    contentWindow:any;
    name:string;

    static staticMethods:string[]=[
        'create',
        // 'getCurrent',
        'wrap',
    ];

    static staticSyncMethods:string[]=[
        'getCurrentSync',
        'wrapSync'
    ];

    static instanceMethods:string[]=[

        // 'getParentApplication',

        'addListener',
        'animate',
        'authenticate',
        'blur',
        'bringToFront',
        // 'close',
        'disableUserMovement',
        'enableUserMovement',
        'executeJavaScript',
        'flash',
        'focus',
        'getAllFrames',
        // 'getBounds',
        'getGroup',
        'getInfo',
        'getNativeId',
        'getOptions',
        'getParentApplication',
        // 'getParentWindow',
        'getState',
        'getWebWindow',
        'getZoomLevel',
        'hide',
        'isMainWindow',
        'isShowing',
        'joinGroup',
        'leaveGroup',
        'maximize',
        'mergeGroups',
        'minimize',
        'moveBy',
        'moveTo',
        'navigate',
        'navigateBack',
        'on',
        'once',
        'prependListener',
        'prependOnceListener',
        'reload',
        'removeAllListeners',
        'removeListener',
        'resizeBy',
        'resizeTo',
        'restore',
        'setAsForeground',
        'setBounds',
        'setZoomLevel',
        'show',
        'showAt',
        'showDeveloperTools',
        'stopFlashing',
        'stopNavigation',
        'updateOptions',
    ];

    static staticInstanceMethods:string[]=[
        'navigateForward',
    ];


    static getCurrent(){
        if (this._name){
            return windowRegistry[this._name];
        }else{
            Window._name='current-window-name';
            return new Window({name:Window._name},()=>{});
        }
    }


    constructor(data:WindowOptions, callback?:Function) {
        super();
        let opened;
        if (data && data.url){
            opened = window.open(data.url,'_blank');
        }
        this.contentWindow = opened || window;
        this.name = data.name;
        if (callback){
            setTimeout(callback,0);
        }
        windowRegistry[data.name] = this;
    }

    close(force?:boolean, callback?:Function,errorCallback?:Function){
        try{
            this.contentWindow.close();
            if(callback){
                callback();
            }
        }catch(error){
            if (errorCallback){
                errorCallback(error);
            }
        }
    }

    getParentWindow(){
        if (window.location.pathname==='/'){
            return this;
        }
        return null;
    }

    getBounds(success:(bounds:WindowBounds)=>void){

        let result:any = {};

        if (success){
            result.left = this.contentWindow.screenX;
            result.top = this.contentWindow.screenY;
            result.width = this.contentWindow.outerWidth;
            result.height = this.contentWindow.outerHeight;

            success(result);
        }
    }

}