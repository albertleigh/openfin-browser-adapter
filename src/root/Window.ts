import {BaseApiClass} from './Base';

import windowRegistry from '../helper/windowRegistry';

declare const window:any;

export interface WindowOptions {
    url?:string,
    name?:string,
}

export interface WindowBounds {
    width:number,
    height:number,
    top:number,
    left:number,
}

export class Window extends BaseApiClass{

    static _name:string = null;

    contentWindow:any;
    identity:{
        uuid:string,
        name:string,
    };

    static staticMethods:string[]=[
        // 'create',
        // 'getCurrent',
        'wrap',
    ];

    static staticSyncMethods:string[]=[
        'getCurrentSync',
        'wrapSync'
    ];

    static instanceMethods:string[]=[

        // 'getParentApplication',

        'addListener',
        'animate',
        'authenticate',
        'blur',
        'bringToFront',
        // 'close',
        'disableUserMovement',
        'enableUserMovement',
        'executeJavaScript',
        'flash',
        'focus',
        'getAllFrames',
        // 'getBounds',
        'getGroup',
        'getInfo',
        'getNativeId',
        'getOptions',
        'getParentApplication',
        // 'getParentWindow',
        'getState',
        'getWebWindow',
        'getZoomLevel',
        'hide',
        'isMainWindow',
        'isShowing',
        'joinGroup',
        'leaveGroup',
        'maximize',
        'mergeGroups',
        'minimize',
        'moveBy',
        'moveTo',
        'navigate',
        'navigateBack',
        'navigateForward',
        'on',
        'once',
        'prependListener',
        'prependOnceListener',
        'reload',
        'removeAllListeners',
        'removeListener',
        'resizeBy',
        'resizeTo',
        'restore',
        'setAsForeground',
        'setBounds',
        'setZoomLevel',
        'show',
        'showAt',
        'showDeveloperTools',
        'stopFlashing',
        'stopNavigation',
        'updateOptions',
    ];


    static async getCurrent():Promise<Window>{
        if (this._name){
            return windowRegistry[this._name];
        }else{
            Window._name='current-window-name';
            return new Window({name:Window._name});
        }
    }


    static getCurrentSync(){
        if (this._name){
            return windowRegistry[this._name];
        }else{
            Window._name='current-window-name';
            return new Window({name:Window._name});
        }
    }


    constructor(data:WindowOptions) {
        super();
        let opened;
        if (data && data.url){
            opened = window.open(data.url,'_blank');
        }
        this.contentWindow = opened || window;
        this.identity.uuid = window.name;
        this.identity.name = data.name;
        windowRegistry[data.name] = this;
    }

    static async create(data:WindowOptions):Promise<Window>{
        return new Window(data);
    }

    async close(force?:boolean):Promise<void>{
        this.contentWindow.close();
        return;
    }

    async getParentWindow():Promise<Window>{
        if (window.location.pathname==='/'){
            return this;
        }
    }

    async getBounds():Promise<WindowBounds>{

        let result:any = {};
        result.left = this.contentWindow.screenX;
        result.top = this.contentWindow.screenY;
        result.width = this.contentWindow.outerWidth;
        result.height = this.contentWindow.outerHeight;

        return result;
    }

}