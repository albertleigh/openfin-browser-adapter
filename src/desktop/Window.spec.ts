import { Window } from './Window';

describe('Desktop::Window',()=>{

    let clazz:typeof Window;

    beforeEach(()=>{
        clazz = Window;
    });

    it('all unsupported static methods defined',()=>{
        expect(clazz.staticMethods).toMatchSnapshot();
    })

    it('all unsupported instance methods defined',()=>{
        expect(clazz.instanceMethods).toMatchSnapshot();
    })

});