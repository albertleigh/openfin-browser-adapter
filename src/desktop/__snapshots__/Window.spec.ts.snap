// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Desktop::Window all unsupported instance methods defined 1`] = `
Array [
  "getNativeWindow",
  "getParentApplication",
  "addEventListener",
  "animate",
  "authenticate",
  "blur",
  "bringToFront",
  "disableFrame",
  "enableFrame",
  "executeJavaScript",
  "flash",
  "focus",
  "getAllFrames",
  "getGroup",
  "getInfo",
  "getOptions",
  "getSnapshot",
  "getState",
  "getZoomLevel",
  "hide",
  "isShowing",
  "joinGroup",
  "leaveGroup",
  "maximize",
  "mergeGroups",
  "minimize",
  "moveBy",
  "moveTo",
  "navigate",
  "navigateBack",
  "navigateForward",
  "reload",
  "removeEventListener",
  "resizeBy",
  "resizeTo",
  "restore",
  "setAsForeground",
  "setBounds",
  "setZoomLevel",
  "show",
  "showAt",
  "stopFlashing",
  "stopNavigation",
  "updateOptions",
]
`;

exports[`Desktop::Window all unsupported static methods defined 1`] = `
Array [
  "wrap",
]
`;
